lib/train/platforms/detect/scanner.rb

```diff
@@ -0,0 +1,84 @@
+# encoding: utf-8
+
+require 'train/platforms/detect/os_common'
+
+module Train::Platforms::Detect
+  class Scanner
+    include Train::Platforms::Detect::OSCommon
+
+    def initialize(backend)
+      @backend = backend
+      @platform = {}
+      @family_hierarchy = []
+
+      # detect cache variables
+      @files = {}
+      @uname = {}
+      @lsb = {}
+    end
+
+    # Main detect method to scan all platforms for a match
+    #
+    # @return Train::Platform instance or error if none found
+    def scan
+      # start with the platform/families who have no families (the top levels)
+      top = Train::Platforms.top_platforms
+      top.each do |_name, plat|
+        next unless plat.detect
```

Comment:
I have added the false lambda instead of the check for detect. This way everything can be evaluated accordingly. 
As far as moving away from instance_eval/exec. The reason we are doing this is so the detect logic will have access to all the helper methods (os_common, os_linux, etc...) in the detect context. We may be able to require all this logic in each instance but that seems overkill for the memory and really doing the same thing under the hood. This also makes it easy to create custom platforms on the fly via inspec or other instances. After talking it over and due to its controlled nature we are voting to keep it for now. I can comment it up to explain a bit more in code.